File: 2015/2/2.jl
part1(filename::AbstractString) = sum(area(line) for line in eachline(filename))

function area(line::AbstractString)
    x, y, z = parse.(Int, split(line, 'x'))
    sides = [x * y, y * z, x * z]
    2 * sum(sides) + minimum(sides)
end

@assert area("2x3x4") == 58
@assert area("1x1x10") == 43

part2(filename::AbstractString) = sum(ribbonlength(line) for line in eachline(filename))

function ribbonlength(line::AbstractString)
    dims = parse.(Int, split(line, 'x'))
    2 * sum(dims) - 2 * maximum(dims) + prod(dims)
end

@assert ribbonlength("2x3x4") == 34
@assert ribbonlength("1x1x10") == 14

File: 2015/2/2.pl
#!/usr/bin/env swipl

:- use_module(library(clpfd)).
:- use_module(library(dcg/basics), [integer//1, blanks//0]).
:- use_module(library(dcg/high_order), [sequence//2]).
:- initialization(main, main).

area(Area) -->
    integer(Length), "x", integer(Width), "x", integer(Height),
    {
        Sides = [Length * Width, Width * Height, Height * Length],
        sum_list(Sides, SumSides),
        min_list(Sides, MinSide),
        Area #= 2 * SumSides + MinSide
    },
    blanks.

:- begin_tests(area).
    :- set_prolog_flag(double_quotes, codes).
    test(example1) :-
        phrase(area(Area), "2x3x4"),
        assertion(Area #= 58).
    test(example2) :-
        phrase(area(Area), "1x1x10"),
        assertion(Area #= 43).
:- end_tests(area).

ribbon_length(RibbonLength) -->
    integer(Length), "x", integer(Width), "x", integer(Height),
    {
        msort([Length, Width, Height], [Small, Medium, _]),
        RibbonLength #= 2 * (Small + Medium) + Length * Width * Height
    },
    blanks.

:- begin_tests(ribbon_length).
    :- set_prolog_flag(double_quotes, codes).
    test(example1) :-
        phrase(ribbon_length(RibbonLength), "2x3x4"),
        assertion(RibbonLength #= 34).
    test(example2) :-
        phrase(ribbon_length(RibbonLength), "1x1x10"),
        assertion(RibbonLength #= 14).
:- end_tests(ribbon_length).

solve_for(DCG, Filename) :-
    once(phrase_from_file(sequence(DCG, Values), Filename)),
    sum_list(Values, Total),
    writeln(Total).

main([]) :-
    main(["input.txt"]).

main([Filename]) :-
    solve_for(area, Filename),
    solve_for(ribbon_length, Filename).
